### PR TITLE
Draft: code reformatting

### PR DESCRIPTION
**Pull request by noc0lour**
This GREP shows ways to implement source code formatting in the GNU Radio source tree and proposes the most sensible way for implementation.

I'd like to encourage discussion about the topic and also get some names of who is interested on nailing down the details of formatting and tooling around it.

### CODE REVIEW

**Review by mbr0wn — COMMENTED** [review]
I would like to suggest my astylerc as a basis for the clang format file.

And I am a fan of method 3 - as long as it's not too far out.

**Comment by noc0lour** [comment]
I'd like to encourage reading or taking a quick glimpse at https://engineering.mongodb.com/post/succeeding-with-clangformat-part-1-pitfalls-and-planning/ as it provides some insights how to pan out reformatting if we want to do it.

File: grep-0002-code-reformatting.md [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+### Conclusion
+
+This GREP proposes to perform a full codebase reformatting according to standards set in GREP 0001 and by specifiying and documenting further formatting decisions in a `.clang-format` in the source tree. Until the reformatting can be implemented several major issues have to be resolved:
+ - Provide sufficient tools for contributors and developers to format new patches and additions according to the GNU Radio coding standards
+ - Prepare code reformatting carefully by writing wrapper around tools to format all parts of the source tree without destroying intentionally formatted parts (Flow charts in comments etc.).
+ - Pick a time of low development activity to push formatting changes on all development branches
+
+To have sufficient diversity in the discussion about crafting a `.clang-format` at least 3 developers should be part of creating it and further tools to format the GNU Radio source tree.
+
+Further on each pull-request and patch as to be formatted using the tools in the GNU Radio source tree before it can be merged.
```
Comment (mbr0wn) [inline grep-0002-code-reformatting.md]: Typo: "has"

**Comment by mbr0wn** [comment]
There could be a hybrid approach: Until we get around to doing method 3, people are allowed to reformat files individually (method 2).

**Comment by carlesfernandez** [comment]
My five cents: at [GNSS-SDR](https://github.com/gnss-sdr/gnss-sdr) (check the `next` branch) we have a `.clang-format` file and guidelines on [how to use it](http://gnss-sdr.org/coding-style/#use-tools-for-automated-code-formatting).

**Comment by marcusmueller** [comment]
I'd also like to point to [`clang-tidy`](http://clang.llvm.org/extra/clang-tidy/), which is awesome in that it enforces for example `{}` around single-line statement blocks (e.g. after `if`, or the `for`) body; for examples, and CMake Integration, see http://www.mariobadr.com/using-clang-tidy-with-cmake-36.html

For example, after I ran CMake with ` -DCMAKE_EXPORT_COMPILE_COMMANDS=ON`, I got a compile_options.json in my build dir. With that,

    /path/to/gnuradio/gr-blocks> clang-tidy  -checks=modernize-use-auto,llvm-include-order,readability-braces-around-statements -fix -p ../build  $(git ls-files|grep '\.cc$')

replaced the redundant explicit type after explicit casts (mostly things like `const gr_complex *in = …` in `work`) with `auto` (which really reduces clutter!), reorders the includes in accordance to what's in GREP1, and adds `{}` around single-line code blocks (`for`, `if`, …).

**Comment by noc0lour** [comment]
I took the python script from the mongodb guys and I'm fiddling with it. There are some real use cases for further development.

**Comment by noc0lour** [comment]
To state my opinion in the PR itself as well: I'm in favour of executing option 3. Since it will allow us to keep the commit noise w.r.t formatting at a minimum (one commit). Traversing through history won't be hindered significantly since only one commit has to be skipped to find previous changes again. And most importantly: we can enforce and keep the code formatted further on using automated tools.

**Comment by marcusmueller** [comment]
@noc0lour would you like to be the Champion for this GREP yourself, or would you rather I or someone else take that job?

**Comment by noc0lour** [comment]
I can do that. I'm in the process of figuring out how we can implement that in https://github.com/gnuradio/gnuradio/pull/1663. I'm open for suggestions and comments in both PRs right now.

File: grep-0002-code-reformatting.md [hunk 2]
```diff
@@ -0,0 +1,67 @@
+# GREP 0002 -- Code reformatting
+
+- Original Author: Andrej Rode <mail@andrejro.de>
+- Status: Draft
```
Comment (marcusmueller) [inline grep-0002-code-reformatting.md:4]: + Champion: The Rode himself.

**Comment by mbr0wn** [comment]
Manually merged: 87b6577ae6351e2ac4a25d5d698f4e204c9ec7c1

**Pull request closed without merging**